src/Lib.P2P/Transports/Udp.cs

```diff
@@ -0,0 +1,135 @@
+    /// </summary>
+    public class Udp : IPeerTransport
+    {
+        private static ILog _log = LogManager.GetLogger(typeof(Udp));
+
+        /// <inheritdoc />
+        public async Task<Stream> ConnectAsync(MultiAddress address,
+            CancellationToken cancel = default)
+        {
+            var port = address.Protocols
+               .Where(p => p.Name == "udp")
+               .Select(p => int.Parse(p.Value))
+               .First();
+            var ip = address.Protocols
+               .Where(p => p.Name == "ip4" || p.Name == "ip6")
+               .First();
+            var socket = new Socket(
+                ip.Name == "ip4" ? AddressFamily.InterNetwork : AddressFamily.InterNetworkV6,
+                SocketType.Dgram,
+                ProtocolType.Udp);
+
+            // Handle cancellation of the connect attempt
+            cancel.Register(() =>
+            {
+                socket.Dispose();
+                socket = null;
+            });
+
+            try
+            {
+                _log.Debug("connecting to " + address);
+                await socket.ConnectAsync(ip.Value, port).ConfigureAwait(false);
+                _log.Debug("connected " + address);
+            }
+            catch (Exception) when (cancel.IsCancellationRequested)
+            {
+                // eat it, the caller has cancelled and doesn't care.
+            }
+            catch (Exception e)
+            {
+                _log.Warn("failed " + address, e);
+                throw;
+            }
+
+            if (cancel.IsCancellationRequested)
+            {
+                _log.Debug("cancel " + address);
+                socket?.Dispose();
+                cancel.ThrowIfCancellationRequested();
+            }
+
+            return new DatagramStream(socket, true);
+        }
+
+        /// <inheritdoc />
+        public MultiAddress Listen(MultiAddress address,
+            Action<Stream, MultiAddress, MultiAddress> handler,
+            CancellationToken cancel)
+        {
+            var port = address.Protocols
+               .Where(p => p.Name == "udp")
+               .Select(p => int.Parse(p.Value))
+               .FirstOrDefault();
+            var ip = address.Protocols
+               .Where(p => p.Name == "ip4" || p.Name == "ip6")
+               .First();
+            var ipAddress = IPAddress.Parse(ip.Value);
+            var endPoint = new IPEndPoint(ipAddress, port);
+            var socket = new Socket(
+                endPoint.AddressFamily,
+                SocketType.Dgram,
+                ProtocolType.Udp);
+            socket.Bind(endPoint);
+
+            // If no port specified, then add it.
+            var actualPort = ((IPEndPoint) socket.LocalEndPoint).Port;
+            if (port != actualPort)
+            {
+                address = address.Clone();
+                var protocol = address.Protocols.FirstOrDefault(p => p.Name == "udp");
+                if (protocol != null)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)